Detect reused keys

andrewRichardson opened the pull request:
### Why am I submitting this PR

Allow detect key reuse and option to warn or fail on reuse.
Example: dev writes `t('foobar')` in two different files ultimately writing translations to the same JSON.
With config `warnOnDuplicates: true`, parser now warns appropriately, and counts the number of warnings like so: `. Warnings:   XXX warnings encountered` to give a general idea of the number of reused keys.
With config `failOnWarnings`, this situation would cause an exit 1.

### Does it fix an existing ticket?

Yes #495 

### Checklist

- [x] only relevant code is changed (make a diff before you submit the PR)
- [x] tests are included and pass: `yarn test` (see [details here](https://github.com/i18next/i18next-parser/blob/master/docs/development.md#tests))
- [x] documentation is changed or added

andrewRichardson commented:
@karellm

karellm commented on bin/cli.js:
```diff
@@ -156,11 +157,30 @@ import i18nTransform from '../dist/transform.js'
             console.log()
             console.log('  Stats:  '.cyan + count + ' files were parsed')
+            if (config.warnOnDuplicates && Object.keys(warningMap).length > 0) {
+              var keys = []
```
I know there are inconsistencies in this file but I'd prefer we stick with `let/const`. Any reason you went for `var`?

andrewRichardson replied:
I went with `var` to match the `count` variable above it, but I can just switch both to `let` to fall more inline with the latest JS practices 👍🏼

karellm commented on bin/cli.js:
```diff
@@ -129,6 +129,7 @@ import i18nTransform from '../dist/transform.js'
   }
 
   var count = 0
+  let warningMap = {}
```
Since this is only used for duplicated keys I'd probably name this `duplicatedKeysCounts` or something similar.

karellm commented on bin/cli.js:
```diff
@@ -156,11 +157,30 @@ import i18nTransform from '../dist/transform.js'
+              keys.sort(function (a, b) {
+                return b[1] - a[1]
+              })
+              console.log(keys)
```
This would warrant a test imo

karellm reviewed (commented):
Thanks for the PR, it looks good! I've added couple comments. My main concern is that tests are a bit light.

andrewRichardson commented:
> @karellm  Thanks for the PR, it looks good! I've added couple comments. My main concern is that tests are a bit light.

I addressed your concerns including adding stronger tests.  Let me know if you have any other concerns! Thanks.

karellm commented on test/cli/cli.test.js:
```diff
@@ -82,6 +82,12 @@ describe('CLI', function () {
+  it('works with `--get-key-duplicates` option', async () => {
+    const subprocess = await execaCommand('yarn test:cli --get-key-duplicates')
+
+    assert.include(subprocess.stdout, "[ [ 'seventh', 3 ] ]")
```
I'm confused why this would be `3` and not `2`. The file has only two occurrences of the key.

andrewRichardson replied:
I went back and looked at how I counted the keys in my own personal implementation compared to the test translation file, and they are slightly different.  That meant that I was adding an extra "count" to each duplicate.  I fixed that in the cli function and it should be sensical now as it correctly counts `2` here etc.

karellm commented on test/helpers/dotPathToHash.test.js:
```diff
@@ -96,6 +96,18 @@ describe('dotPathToHash helper function', () => {
+    const { target, duplicate, conflict } = dotPathToHash(
+      { keyWithNamespace: 'one.two.three' },
+      { one: { two: { three: '' } } },
+      { warnOnDuplicate: true }
```
The option is `warnOnKeyDuplicate` and not `warnOnDuplicate`. There is something flawed with the logic in this PR. This option shouldn't be needed to be passed to `dotPathToHash`. It should already return duplicates and conflicts without any extra option. This test is basically a duplicate of the one right above it.

andrewRichardson replied:
Addressed in your other comment below

karellm replied:
@andrewRichardson Is there a reason you left the test? It doesn't seem to test anything since you've removed `warnOnDuplicate`.

andrewRichardson replied:
Ah, you are correct.  This was left by mistake, I just pushed a commit to remove it 👍🏼

karellm commented on src/transform.js:
```diff
@@ -182,7 +185,7 @@ export default class i18nTransform extends Transform {
             )
           } else if (conflict === 'value') {
-            this.warn(`Found same keys with different values: ${entry.key}`)
+            this.warn(KEY_REUSE_WARNING + entry.key)
```
I believe what we want is a new warning for cases where `duplicate === true` and `conflict === false` which currently is not covered by the code. You should rename `KEY_REUSE_WARNING` to `KEY_CONFLICT_WARNING` and create a new warning for `KEY_REUSE_WARNING`.

andrewRichardson replied:
Understood!  It seems like I ended up targeting only my use case and missed the bigger picture.  I reworked the code to use the `conflict = false` condition and everything seems to be working well, great suggestion.

karellm reviewed (commented):
Doing a last review I noticed inconsistencies with the code. Please review my two comments. I think the approach to solve this issue is not ideal and I've suggested an alternative.

andrewRichardson commented:
@karellm bump

amandapouget commented:
This feature seems really cool! Can we get to merge on it?

andrewRichardson commented:
@karellm bump x2

andrewRichardson commented:
@karellm bump x3

andrewRichardson commented:
@karellm bump x4?
